File: BOJ/15649.py
import sys

def choose(set_nums):
    global answers,n,m
    if len(set_nums) == m:
        answers.append(set_nums)
    else:
        for i in range(1,n+1):
            if checked[i]:continue
            checked[i] = True
            set_nums.append(i)
            choose(set_nums.copy())
            checked[i] = False
            set_nums.remove(i)

input = sys.stdin.readline

n, m = map(int,input().split())

set_nums = []
checked = [False for _ in range(n+1)]
answers = []
choose(set_nums)

for answer in answers:
    for a in answer:
        print(a,end=" ")
    print()

File: BOJ/15649.kt
import java.io.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, m) = br.readLine().split(' ').map { it.toInt() }
    val checked = BooleanArray(n+1,{false})
    val set_nums = mutableListOf<Int>()
    choose(set_nums,n,m,checked)

}
fun choose(set_nums:MutableList<Int>,n:Int,m:Int,checked:BooleanArray){
    if (set_nums.size == m){
        for(num in set_nums){
            print("$num ")
        }
        println()
    }
    else{
        for (num in 1..n){
            if (checked[num]) continue
            checked[num] = true
            set_nums.add(num)
            choose(set_nums,n,m,checked)
            checked[num] = false
            set_nums.remove(num)
        }
    }
}

